### DISCUSSION TITLE
Dfference between yabai's space rule and dock's assign to desktop

### DISCUSSION THREAD

**Discussion by szfkamil** [asmvik/yabai#2226] [General]
There exists overlap between native OS functionality and yabai. Specifically, I'd like to touch on the space rule, as well as the assign to space option from Dock. 

The `space` rule simply sends the window to space. 

`Send window to space. If ^ is present, follow focus.`

The assign to space function in dock seems to always open the window in that space. What's the difference between both functionalities? Is it a bad idea to have both enabled per app?

Should space assignments be done via the dock or yabai. Why, why not?

**Comment by szfkamil** [comment]
I have decided to investigate the behaviour myself and have some notes that might be helpful for future readers. Please feel free to point out any mistakes and make suggestions.

If window gets opened and has space rule assigned with yabai, it briefly flashes on current space and gets sent to the assigned space (would be nice to know how to prevent this flash)

If a window is assigned to space via the dock.app, and you have the option “When switching to an application, switch to Space with open windows for the application” under Desktop & Dock enabled, it will open on the assigned space with focus.

If a window is assigned to the same space per dock and yabai, they will not conflict. However, I have not investigated the case where space rule and space assignment differ. In this case, the window will not 'flash' on current space and will behave as if the yabai space rule were not there at all.

The only practical difference between yabai and dock, is that yabai allows per-window assignment with RegEx, whereas dock.app assigns apps per space. 

Note: Perhaps it is worth investigating what happens if you have different space rules for windows of a specific application and dock assignments for that app.

Apart from that, having a structured `yabairc` helps you see in plain text what app/window is assigned to what space, whereas with the dock assignments, you can only see per-app assignment via right-clicking on the app icon in the dock. 


Furthermore, with yabai, the windows briefly flash and do not focus the window once it is opened. If you are not on the space the window is assigned to, you have to go to that space first and focus the app, or 'launch' it the second time (if you have “When switching to an application, switch to Space with open windows for the application” enabled)

This is just my investigation of the behaviour on yabai v7.1.0

**Reply by mattiacerutti** [reply]
Hello,

I know this is a year-old discussion, but i was wondering if you ever ended up figuring out how to prevent the flash you mentioned when you said:

> If window gets opened and has space rule assigned with yabai, it briefly flashes on current space and gets sent to the assigned space (would be nice to know how to prevent this flash)

Its driving me crazy.

**Reply by szfkamil** [reply]
I don't remember. Whatever I did figure out is attached in my comment. I don't use these features of yabai anymore.

**Reply by szfkamil** [reply]
> Hello,
> 
> I know this is a year-old discussion, but i was wondering if you ever ended up figuring out how to prevent the flash you mentioned when you said:
> 
> > If window gets opened and has space rule assigned with yabai, it briefly flashes on current space and gets sent to the assigned space (would be nice to know how to prevent this flash)
> 
> Its driving me crazy.

It's not really a bug, so it can't be 'fixed'. 

What just popped to mind is trying to script with jq to change the window opacify to 0 (basically not visible), set a sleep timer for like 2 seconds, and change opacify back to the value before setting it to 0. 

You would store the value of opacity before the change, then set it as a new opacity after the sleep timer ends.

If you have no idea what I'm saying I can't help. I had this pop to my mind just now. Good luck!

**Reply by mattiacerutti** [reply]
That was my original idea too! 

I tried setting up a yabai signal for the event `application_launched` but i unfortunately came to the conclusion that this event (alongside with all the others) is triggered just moments **AFTER** a window becomes visible on my screen (or maybe yabai is too slow at reacting, idk).

This kind of limitations prevents me to do any kind of shenanigans because i don't have something to react to in time.

**Reply by asmvik** [reply]
Yabai cannot preempt the macOS compositor. We are only notified that the window exists after it has been displayed on the screen.

**Reply by mattiacerutti** [reply]
> Yabai cannot preempt the macOS compositor. We are only notified that the window exists after it has been displayed on the screen.

Another thing i've noticed that contributes to this "flashing" visual effect is the following.

Let's say i have: 

```yaml
yabai -m rule --add app="Foo" space=^2 manage=off
yabai -m rule --add app="Fee" space=^2 manage=off

```

When opening `Fee`, its window gets correctly sent to `space 2` and the focus follows. 
If i switch back to `space 1` and open `Foo`, its window gets sent to `space 2` and the focus follows, but `Foo's` window appears **BELOW** the previously opened `Fee's` window — requiring a re-focus to make it appear.

This behavior does not happen if, before opening `Foo` i do not switch to another space (in this case space 1).
